### DISCUSSION TITLE
PIP-6: In-Memory Storage of Public Keys, Validators, and Accounts using LRU Cache

### DISCUSSION THREAD

**Discussion by ja7ad** [pactus-project/PIPs#17] [PIPs]
PIP-6: [In-Memory Storage of Public Keys, Validators, and Accounts using LRU Cache](https://pips.pactus.org/PIPs/pip-6)

**Comment by b00f** [comment]
I have a few suggestions:

1- Let's change the title to "In-Memory Caching for stored items". This is broader, and can cover more than just Validators, Accounts, and Public keys, like transactions.

2- We don't always need to use LRU. Depending on the situation, we might choose a different data structure. For instance, to keep a list of the recent transactions in the blockchain, we might use a [deque](https://en.wikipedia.org/wiki/Double-ended_queue). In each block we can add new transactions at the head the deque and remove older ones from the tail. This helps improve [PIP-2](https://pips.pactus.org/PIPs/pip-2).

3- This statement isn't quite right:
```
Efficiently managing public keys, validators, and account data is crucial for reducing blockchain storage requirements and enhancing overall system performance.
```
A cache doesn't "reducing blockchain storage requirements" but it does "enhancing overall system performance".

4- Actually, LevelDB already has an LRU cache. But it's not very efficient since it only stores raw data. This raw data then needs to be converted to the right structure (like Account). An in-house cache avoids this, so we should update the Specification accordingly.

**Reply by ja7ad** [reply]
@allahbakhsh @parsa-hedayatnia what do you think about these suggestions?

**Reply by parsa-hedayatnia** [reply]
1- Changing the name to In-Memory Caching for stored items is a good idea, but storing more data than validations, accounts and public keys, except for the mentioned ones, needs to be checked more because we have to check whether the number of times used in the intervals Time is so valuable that we store them in our LUR.
2- In my opinion, LUR is a more appropriate option than the proposed one, because according to what I have studied, deque is a general purpose data structure that is used to add and remove data from both ends of a queue. But a LUR is a specialized data structure used to store and ensure that frequently accessed items are readily available.
3- For the third part, it's true. The sentence is not completely true. The cache does not directly reduce the storage of the blockchain, but it significantly improves the overall process of the system. For different items, this improvement will be different according to the frequency of use. .
4- The problem raised is absolutely true that there are several solutions to solve it
The first solution: Custom Cache structure that we can refer to LevelDB source code leveldb/util/cache.cc to know how to implement it.
Second solution: Advanced storage technique is to explore layered storage, for example, where we have multiple cache levels with different data processing levels, which can help optimize the conversion process for different data types, in line with the original proposal. It is also more effective to use other data besides validations, accounts and public keys.
Third solution: I'm not entirely sure, but I think hybrid data storage is an approach that can strike a balance between storage efficiency and data retrieval performance that needs further investigation.

In the end, it should be considered that this proposal (Pip-6) will optimize performance and reduce data storage requirements and increase scalability by reducing the load on the underlying data storage system, but at the same time, important issues such as security implications that Some of its cases were mentioned in Pip- 4, concurrency issues that refer to the management of simultaneous access to the cache, and most importantly, resource consumption. A cache in memory consumes system resources, including RAM. Depending on the size of the cache and the amount of data it stores, this can affect the hardware requirements and costs of running a blockchain node, which from what I've read from Pactus should be able to handle 32 validators on a local system. run on a single node, so these conditions should also be considered.

**Comment by b00f** [comment]
@parsa-hedayatnia 

For special cases of transactions, we should ensure that we keep **recent** transactions in the cache. By "recent," we mean transactions from the current block back to [Time-To-Live interval](https://pactus.org/learn/consensus/parameters/) blocks. 

```
- [ ] - [ ] - [ ] - [ ] .... [ ] - [ ]
         |<-------------------------| 
                     TTL
```
The reason is that we want to ensure the transaction hasn't been committed previously, as discussed in [PIP-3](https://pips.pactus.org/PIPs/pip-3).

Regarding the various solutions, you have a valid point. We have multiple layers of caching, and as you mentioned, we should define the size of the cache carefully for optimal resource management.

About the number of validators, even though each node can support a maximum of 32 validators, we maintain just one store for all of them. So, we don't duplicate storage for each validator, and there's no need to be concerned about it.
